### PR TITLE
Fix message encoding and encryption.

### LINKED ISSUE #181: Test ably-common/test-resources/crypto-data-*.json.
Objective-C tests were removed #171. They were kind of hacky anyway and can't be trivially translated to the new design. They need to be redone in Swift.

### LINKED ISSUE #195: ARTDataDecoded: error decoding a NSDictionary or NSArray
`RSL4d4` is showing:

```
2016-02-04 19:03:17.233 xctest[85570:13716774] ERROR: ARTDataDecoded failed to decode data as 'json': (__NSCFArray)(
    John,
    Mary
)
2016-02-04 19:03:17.233 xctest[85570:13716774] ERROR: ARTDataDecoded failed to decode data as 'json': (__NSCFDictionary){
    name = John;
    number = 3;
}
```

### PR DESCRIPTION
**Pull request by tcard**
Now testing with fixtures in ably-common.

Fixes #181.
Fixes #195.

### CODE REVIEW

File: ablySpec/Crypto.swift [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+//
+
+import Nimble
+import Quick
+import SwiftyJSON
+
+class Crypto : QuickSpec {
+    override func spec() {
+        describe("Crypto") {
+            for keyLength in ["128", "256"] {
```
Comment (ricardopereira) [inline ablySpec/Crypto.swift]: Did something similar: https://github.com/ably/ably-ios/pull/209/files#diff-ef1119dd0f908b6d236fe852b7c431f7R152 for #209

Comment (tcard) [inline ablySpec/Crypto.swift]: OK, once #209 is merged I'll use that function. LGTY otherwise?

**Comment by mattheworiordan** [comment]
Looks good, although as the Encoder & Decoder are being used directly, would it be worth adding at least one encryption & decryption test to ensure it works end to end, or do we already have that?

**Comment by tcard** [comment]
@mattheworiordan You mean, send and receive through an encrypted channel? I think #209 which tests [RTL6i](http://docs.ably.io/client-lib-development-guide/features/#RTL6i) takes/should take care of that.

**Comment by mattheworiordan** [comment]
> I think #209 which tests RTL6i takes/should take care of that.

True, so then :+1:

File: ably-ios/ARTBaseMessage.m [hunk 2]
```diff
@@ -40,20 +40,26 @@ - (instancetype)messageWithData:(id)data encoding:(NSString *)encoding {
...
 
-- (ARTStatus *)decodeWithEncoder:(ARTDataEncoder*)encoder output:(id *)output {
+- (id)decodeWithEncoder:(ARTDataEncoder*)encoder error:(NSError **)error {
     ARTDataEncoderOutput *decoded = [encoder decode:self.data encoding:self.encoding];
-    *output = [self copy];
-    ((ARTBaseMessage *)*output).data = decoded.data;
-    ((ARTBaseMessage *)*output).encoding = decoded.encoding;
-    return decoded.status;
+    if (decoded.status.state != ARTStateOk && error) {
+        *error = [NSError errorWithDomain:ARTAblyErrorDomain code:0 userInfo:@{@"reason": @"decoding failed"}];
```
Comment (ricardopereira) [inline ably-ios/ARTBaseMessage.m]: Please use `NSLocalizedDescriptionKey` or `NSLocalizedFailureReasonErrorKey`. It's a convention throughout Cocoa.

 - NSLocalizedDescriptionKey: A localized description of the error.
 - NSLocalizedFailureReasonErrorKey: A localized explanation of the reason for the error.

E.g.:
```
NSDictionary *userInfo = @{ NSLocalizedDescriptionKey: @"Operation was unsuccessful." };
```

File: ably-ios/ARTBaseMessage.m [hunk 3]
```diff
@@ -40,20 +40,26 @@ - (instancetype)messageWithData:(id)data encoding:(NSString *)encoding {
...
 
-- (ARTStatus *)encodeWithEncoder:(ARTDataEncoder*)encoder output:(id *)output {
+- (id)encodeWithEncoder:(ARTDataEncoder*)encoder error:(NSError **)error {
     ARTDataEncoderOutput *encoded = [encoder encode:self.data];
-    *output = [self copy];
-    ((ARTBaseMessage *)*output).data = encoded.data;
-    ((ARTBaseMessage *)*output).encoding = [self.encoding artAddEncoding:encoded.encoding];
-    return encoded.status;
+    if (encoded.status.state != ARTStateOk && error) {
+        *error = [NSError errorWithDomain:ARTAblyErrorDomain code:0 userInfo:@{@"reason": @"encoding failed"}];
```
Comment (ricardopereira) [inline ably-ios/ARTBaseMessage.m]: As above.

File: ably-ios/ARTChannel.m [hunk 4]
```diff
@@ -67,9 +67,10 @@ - (ARTMessage *)encodeMessageIfNeeded:(ARTMessage *)message  {
     if (!self.dataEncoder) {
         return message;
     }
-    ARTStatus *status = [message encodeWithEncoder:self.dataEncoder output:&message];
-    if (status.state != ARTStateOk) {
-        [self.logger error:@"ARTChannel: error encoding data, status: %tu", status];
+    NSError *error = nil;
+    message = [message encodeWithEncoder:self.dataEncoder error:&error];
+    if (error != nil) {
+        [self.logger error:@"ARTChannel: error encoding data"];
```
Comment (ricardopereira) [inline ably-ios/ARTChannel.m]: What do you think about logging the `error`? I think it could help.

File: ably-ios/ARTRealtimeChannel.m [hunk 5]
```diff
@@ -322,9 +322,10 @@ - (void)onMessage:(ARTProtocolMessage *)message {
     for (ARTMessage *m in message.messages) {
         ARTMessage *msg = m;
         if (dataEncoder) {
-            ARTStatus *status = [msg decodeWithEncoder:dataEncoder output:&msg];
-             if (status.state != ARTStateOk) {
-                [self.logger error:@"ARTRealtimeChannel: error decoding data, status: %tu", status];
+            NSError *error = nil;
+            msg = [msg decodeWithEncoder:dataEncoder error:&error];
+            if (error != nil) {
+                [self.logger error:@"ARTRealtimeChannel: error decoding data"];
```
Comment (ricardopereira) [inline ably-ios/ARTRealtimeChannel.m]: Log `error`?

File: ably-ios/ARTRealtimeChannel.m [hunk 6]
```diff
@@ -347,9 +348,10 @@ - (void)onPresence:(ARTProtocolMessage *)message {
     for (ARTPresenceMessage *p in message.presence) {
         ARTPresenceMessage *pm = p;
         if (dataEncoder) {
-            ARTStatus *status = [pm decodeWithEncoder:dataEncoder output:&pm];
-             if (status.state != ARTStateOk) {
-                [self.logger error:@"ARTRealtimeChannel: error decoding data, status: %tu", status];
+            NSError *error = nil;
+            pm = [pm decodeWithEncoder:dataEncoder error:&error];
+             if (error != nil) {
+                [self.logger error:@"ARTRealtimeChannel: error decoding data"];
```
Comment (ricardopereira) [inline ably-ios/ARTRealtimeChannel.m]: Log `error`?

**Comment by ricardopereira** [comment]
Beside my comments, LGTM.

File: ably-ios/ARTChannel.m (hunk 4 above)
Comment (tcard) [inline ably-ios/ARTChannel.m]: Done at dc2cfe8.

File: ably-ios/ARTRealtimeChannel.m (hunk 6 above)
Comment (tcard) [inline ably-ios/ARTRealtimeChannel.m]: Done at 269f07f.

File: ably-ios/ARTRealtimeChannel.m (hunk 5 above)
Comment (tcard) [inline ably-ios/ARTRealtimeChannel.m]: Done at 269f07f.

File: ably-ios/ARTBaseMessage.m (hunk 2 above)
Comment (tcard) [inline ably-ios/ARTBaseMessage.m]: Done at 269f07f.

**Comment by tcard** [comment]
@ricardopereira Can you take a look at 269f07f?

**Comment by ricardopereira** [comment]
Done :+1:

**Pull request merged**